client/mysqlbinlog.cc

```diff
@@ -3490,6 +3491,8 @@ static Exit_status dump_local_log_entries(PRINT_EVENT_INFO *print_event_info,
       /*
         Emit a warning in the event that we finished processing input
         before reaching the boundary indicated by --stop-position.
+        Due to design limitations, Binlog_gtid_state_validator::report
```

Comment:
> *Though I wonder if that is already what you had planned before I interfered in the fist place 😇*

Yes, it was similar.
I originally considered adding an API to `Domain_gtid_event_filter *position_gtid_filter` that exposes the completion states of each domain-specific subfilter.
I was ultimately convinced that hiding the code in `gtid_state_validator` instead is a lesser change to APIs. After all,
> *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

I’d definitely choose the former, more consistent strategy if

> *This is a new feature or a refactoring, and the PR is based against the main branch.*

Reply:
> For more straightforward reading, it'd be better to highlight the inconsistency at the start of the sentence. Also, "due to design limitations" doesn't really provide context into why (which is good to have when going back through time :), so it would be good to expand upon that).

I have my rationals in the commit description. I’ll copy them over.

Reply:
Here’s what we could do: Apply these introverted changes for `10.11` branch, but also follow-up with a refactor (see top post) on the `main` branch.

Reply:
> I originally considered adding an API to Domain_gtid_event_filter *position_gtid_filter that exposes the completion states of each domain-specific subfilter.

You could also do the warnings in the filters. There is already `has_finished()`, which you could extend with an optional err-mode parameter (with some noop default) , and then call `has_finished(<err_mode_config>)` here (or something like that. One idea could be a function-pointer to an output function (like `report_f` in `report_audit_findings` (just an idea, you could do it how you like, if you agree with the overall direction)). _Though you'll also have to extend the filters a little to support this :)_

Reply:
Utilizing `has_finished` requires modifying it for the entire `Gtid_event_filter` family, or at least both the `Domain_gtid_event_filter` and the `Window_gtid_event_filter`s it’s comprised of.
(I don’t know if C++ subclasses can override parent methods with optional params.)

This scale is no less than creating a new method (which is a new feature) outright, which is why I’ve been hesitant to do these in `10.11` but would feel unrestrained on `main`.

Moreover, although the location would be right, the code itself would still be inconsistent with the surrounding #3400 content – this would be `has_finished(/* warnings */ TRUE)` while those were `if (has_finished()) warn("…")`.
Coercing #3400’s logic to be similar to this would somewhat undo them entirely.
In comparison, my original idea looked like:
```c++
for(Window_gtid_event_filter stop_filter: position_gtid_filter.m_stop_filters) {
  if(!stop_filter.has_finished())
    warn("…");
}
```

Reply:
Right, I see your point. Although looking deeper through our current API, we actually already can (basically) do what you suggest, with minimal changes (and through the `state_validator`). We can use `Binlog_gtid_state_validator::verify_gtid_state()` and iterate through our stop-position GTIDs to use as the `domain_state_gtid`, and modify the code so if `out` is `NULL`, then we don't print the warnings within the function, but we just return `FALSE`. So pseudo-codically:

```
rpl_gtid *stop_gtids= position_gtid_filter.get_stop_gtids(); // or maybe to cache them at a previous use?
for(rpl_gtid *stop_gtid: stop_gtids) {
  if(!gtid_state_validator.verify_gtid_state(NULL, stop_gtid))
    warn("…", stop_gtid)
}
my_free(stop_gtids);
```

Reply:
I think we'd be ok extending the filter API in 10.11, so long as we do it minimally (i.e. I'm not a huge fan of exposing the filters directly as `position_gtid_filter.m_stop_filters`). Perhaps we can add a function to `Domain_gtid_event_filter` of signature `rpl_gtid *Domain_gtid_event_filter::get_stop_gtids_for_active_filters(uint n)` which abstracts the loop and `has_finished()` call

Reply:
I have reëvaluated that it’s not worth to implement suboptimally and inconsistentlly (and also tamper the `delete gtid_state_validator` optimization) just to minimalize API differences.

> I think we'd be ok extending the filter API in 10.11, so long as we do it minimally.

I am going to reïmplement this and #3400 with my ideal concept imagined in the top post:
Move the warning feature to within the (entire?) `Gtid_event_filter` family.
This will probably be a new method (rather than repurposing `has_finished`) so it futureproofs for other filters such as `--start-position` and even `--ignore-server-ids` ([MDEV-20119](https://jira.mariadb.org/browse/MDEV-20119)).

If this turns out to be too much change for 10.11, I might rebrand this ticket & PR to a new feature on `main`.
(I won’t refactor the validator, at least for now.)

---

Speaking of – according to `git grep`, this is an unimplemented prototype found only in its header:
https://github.com/MariaDB/server/blob/a37f71bd10fea224b8f198cb42ca0320dffaf8eb/sql/rpl_gtid.h#L666-L674

Reply:
Sounds good (though I could never find what you were referring to with):
> with my ideal concept imagined in the top post:

And please feel free to remove `verify_gtid_is_expected()` while you're at it :) (though as a separate cleanup commit would be best)

Reply:
> with my ideal concept imagined in the top post:

oh, it’s [collapsed](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-collapsed-sections) under “▶ Unlike 3400”